Drop PHP 7.4 support, upgrade to vimeo/psalm v5

internalsystemerror opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

This PR:
- Adds support for PHP 8.2
- Drops support for PHP 7.4
- Updates `vimeo/psalm` to v5
- Solves various psalm errors (but also creates a couple of new ones added to the baseline sadly).

Closes:
- https://github.com/laminas/laminas-cache-storage-adapter-redis/pull/57
- https://github.com/laminas/laminas-cache-storage-adapter-redis/pull/58

Commit:
add PHP 8.2 support, drop PHP 7.4

Signed-off-by: Gary Lockett <gary@creativecow.uk>

Commit:
update vimeo/psalm to v5

Signed-off-by: Gary Lockett <gary@creativecow.uk>

internalsystemerror commented:
There are various errors shown in CI here:
- `boesing/psalm-plugin-stringf` not supporting PHP 8.2 (solved by: https://github.com/boesing/psalm-plugin-stringf/pull/142)
- Too many open files (not sure what is causing this still)
- `RedisCluster::keys(): Can't send KEYS to 172.18.0.3:6380` (which I'm unable to test locally as I don't have redis cluster setup).

Hopefully this can at least be a place to start if anyone else wishes to pick this up before I can fix, but those last 2 are really stumping me.

boesing commented:
I already spent some time in trying to fix CI regarding Open Files and KEYS error.

Will see that the open PRs get merged once I get CI running. I do prefer not merging this without CI passing for the unit tests.


Too many open files is the error I mentioned in slack. Its ridicculous and drives me crazy as all CI errors we have right now are hard to tackle...

internalsystemerror commented:
The open files issue only occurs with the latest dependencies, if I could reproduce that locally (which I know you mentioned you can't either) then we might be able to figure out which version of which dep and conflict with it. Or even open an issue upstream.

boesing commented:
@internalsystemerror I am currently merging the renovate PRs. Are you open to apply your changes to the renovate PR regarding psalm v5?

internalsystemerror commented:
I prefer not to work directly on the renovate branches myself as it prevents renovate from updating them in the future so its a habit I've gotten into. But I can rebase this one using the commit from the renovate branch tomorrow.

renovate[bot] committed:
Update dependency vimeo/psalm to v5

| datasource | package     | from   | to    |
| ---------- | ----------- | ------ | ----- |
| packagist  | vimeo/psalm | 4.30.0 | 5.2.0 |

Signed-off-by: Renovate Bot <bot@renovateapp.com>
Signed-off-by: Gary Lockett <gary@creativecow.uk>

Ocramius commented on src/RedisResourceManager.php:
```diff
@@ -233,9 +233,8 @@ protected function normalizeServer(&$server)
 
-            $host    = $server['host'];
-            $port    = isset($server['port']) ? (int) $server['port'] : $port;
-            $timeout = isset($server['timeout']) ? (int) $server['timeout'] : $timeout;
```
This code block is not unused? 🤔

internalsystemerror replied:
🤷

Ocramius replied:
Why did you drop it then? :D

internalsystemerror replied:
It is actually unused as `$server['timeout']` is never available at this point. To get here `$server` must be a string that is then turned into array via `parse_url` containing no timeout key, or a hard `['host' => ...]` again without any timeout key.

Ocramius replied:
aha, I see: so `$server['timeout']` can never exist here 👍

internalsystemerror replied:
Exactly, sorry I misunderstood your first comment here.

Ocramius commented on src/RedisResourceManager.php:
```diff
@@ -596,7 +595,7 @@ protected function normalizeLibOptionKey(&$key)
             throw new Exception\InvalidArgumentException("Unknown redis option '{$key}' ({$const})");
         }
         /** @var int $key */
-        $key = constant($const);
+        $key = (int) constant($const);
```
IMO a cast is worse here: it obfuscates cases where non-numeric values are fetched

internalsystemerror replied:
All these constant values should be an integer I believe

Ocramius replied:
Then we should at least `assert()` instead :-)

internalsystemerror replied:
Yeah I'd like to do much much more to ensure these types, and I think we may need to just accept BC breaks and a new major to do so.

internalsystemerror replied:
e.g. `normalizePersistentId` method just casts the variable reference to a string and nothing else, when the param type is already set to be a string. So assuming that the types are correct, this method is redundant, but that may actually break usages which aren't currently checking for types.

Ocramius replied:
for now I'd really just recommend coding the assumption in form an assertion though

internalsystemerror replied:
Will do for this line, but isn't that already a BC Break in case any of these aren't an integer?

internalsystemerror commented:
I've rebased but now I'm getting a load of extra psalm errors. I think the options array needs typing and I'm kinda confused what's expected, e.g. `$server['timeout']` is used where `$server['port']` should be, things like that.

Commit:
fix some of the easier psalm errors

Signed-off-by: Gary Lockett <gary@creativecow.uk>

Commit:
use assertion over casting

Signed-off-by: Gary Lockett <gary@creativecow.uk>

internalsystemerror commented:
Yeah, so to be specific. The issue is because in a lot of places rather than throwing exceptions, values are being coerced via normalisation which means to fix most of these issues is a BC break as far as I can see. If using specific types the vast majority of this normalisation becomes redundant also.

Commit:
import is_int function#

Signed-off-by: Gary Lockett <gary@creativecow.uk>

Commit:
add coeerced type errors to baseline

Signed-off-by: Gary Lockett <gary@creativecow.uk>

Ocramius reviewed (approved):
Thanks @internalsystemerror!
